sympy/sets/sets.py

```diff
@@ -1202,28 +1202,32 @@ def reduce(args):
 
         # ===== Pair-wise Rules =====
         # Here we depend on rules built into the constituent sets
-        args = set(args)
+        args = FiniteSet(*args)
         new_args = True
         while(new_args):
             for s in args:
                 new_args = False
-                for t in args - set((s,)):
+                args_without_s = set(args) - set(FiniteSet(s))
+                args_without_s = FiniteSet(*[ar for ar in args_without_s])
+                for t in args_without_s:
                     new_set = s._union(t)
                     # This returns None if s does not know how to intersect
                     # with t. Returns the newly intersected set otherwise
                     if new_set is not None:
                         if not isinstance(new_set, set):
                             new_set = set((new_set, ))
-                        new_args = (args - set((s, t))).union(new_set)
+                        remv = set((s, t))
+                        new_args = (set(args) - remv).union(new_set)
                         break
                 if new_args:
+                    new_args = FiniteSet(*[ar for ar in new_args])
                     args = new_args
```

Comment:
@Shekharrajak as I said above the actual issue is here, modifying the list your are iterating. You need to rewrite the whole block so that this doesn't happen.